### PR TITLE
JS: Support import attributes

### PR DESCRIPTION
**Pull request by aibaars**
Adds support for [import attributes](https://github.com/tc39/proposal-import-attributes), an evolution of [import assertions](https://github.com/tc39/proposal-import-assertions). For example:
```js
import * as foo from 'foo' with { type: 'json' }
```

When implementing https://github.com/github/codeql/pull/12382, we thought the syntax of import assertions was unlikely to change. However, it did. The only difference is that instead of `assert` they now use `with`. The old syntax is still allowed but deprecated.

[Evaluation](https://github.com/github/codeql-dca-main/issues/16763) was uneventful.

Commit-by-commit review recommended.

### CODE REVIEW

**Comment by erik-krogh** [comment]
If you revert the TypeScript specific changes, then I think we can land this as a separate PR before TypeScript 5.3 is released.

File: javascript/ql/lib/semmle/javascript/Expr.qll [hunk 1]
```diff
@@ -2823,9 +2823,9 @@ class DynamicImportExpr extends @dynamic_import, Expr, Import {
...
    * Gets the second "argument" to the import expression, that is, the `Y` in `import(X, Y)`.
    *
-   * For example, gets the `{ assert: { type: "json" }}` expression in the following:
+   * For example, gets the `{ with: { type: "json" }}` expression in the following:
    * ```js
-   * import('foo', { assert: { type: "json" }})
+   * import('foo', { with: { type: "json" }})
    * ```
    */
   Expr getImportAttributes() { result = this.getChildExpr(1) }
```
Comment (asgerf) [inline javascript/ql/lib/semmle/javascript/Expr.qll:2831]: There's an annoying discrepancy here, in that this doesn't return the same type of thing as for `ImportDeclaration` and `ExportDeclaration`:

```js
// getImportAttributes() = { type: "json" }
import "foo" with { type: "json" }

// getImportAttributes() = { with: { type: "json" } }
import("foo", { with: { type: "json" }} )
```

The enclosing `with` property is omitted in the first version and (by necessity) included in the second version.

That's why this was already called `getImportAttributes` and other one was called `getImportAssertion`, to clarify that they don't do the same thing. The term "import attribute" was initially something I came up with to avoid the ambiguity. Unfortunately, the new proposal _also_ uses the term "import attribute", except with a different meaning. It seems in the new proposal, this value is [called the "options bag"](https://github.com/tc39/proposal-import-attributes#dynamic-import), although I'd suggest a slightly nicer name `getImportOptions()`.

TLDR: Please rename this to `getImportOptions()` and deprecate the old name (as it does not in fact return the import attributes).

**Review by asgerf — COMMENTED** [review]
Looks great overall! 👍

File: javascript/extractor/src/com/semmle/ts/extractor/TypeScriptASTConverter.java [hunk 2]
```diff
@@ -342,10 +343,10 @@ private Node convertNodeUntyped(JsonObject node, String defaultKind) throws Pars
...
       case "AsExpression":
         return convertTypeAssertionExpression(node, loc);
-      case "AssertClause":
-        return convertAssertClause(node, loc);
-      case "AssertEntry":
-        return convertAssertEntry(node, loc);
+      case "ImportAttributes":
+        return convertImportAttributes(node, loc);
+      case "ImportAttribute":
+        return convertImportAttribute(node, loc);
```
[the remark this answers is not in the document]
Comment (aibaars) [reply javascript/extractor/src/com/semmle/ts/extractor/TypeScriptASTConverter.java:349]: It's plural vs singular  ImportAttribute(**s**) .

File: javascript/ql/lib/semmle/javascript/Expr.qll (hunk 1 above)
Comment (aibaars) [reply javascript/ql/lib/semmle/javascript/Expr.qll:2831]: Thanks for spotting that, I hadn't realised that they were actually different. I just assumed you somehow had a crystal ball to foretell the future name of the feature ;-)

Comment (aibaars) [reply javascript/ql/lib/semmle/javascript/Expr.qll:2831]: Can't we somehow make `getImportAttributes()` peel off the `with:` and return `{ type: "json" }` for 
```
import("foo", { with: { type: "json" }} )
```

Although I guess this gets tricky for cases like
```
let options = { with: { type: "json" }}
import("foo", options)
```

Comment (erik-krogh) [reply javascript/ql/lib/semmle/javascript/Expr.qll:2831]: > Can't we somehow make `getImportAttributes()` peel off the `with:` and return `{ type: "json" }` for

I don't think this part of the standard library should do that. I think that should be up the query writers (and they can use API-graphs).

File: javascript/extractor/src/com/semmle/ts/extractor/TypeScriptASTConverter.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (asgerf) [reply javascript/extractor/src/com/semmle/ts/extractor/TypeScriptASTConverter.java:349]: Oh you're right 🙈

File: javascript/ql/lib/semmle/javascript/Expr.qll (hunk 1 above)
Comment (asgerf) [reply javascript/ql/lib/semmle/javascript/Expr.qll:2831]: Agree with @erik-krogh, and in any case we still need `getImportOptions()` for the general case. Also let's not worry too much about building convenience predicates that we might never need.

**Comment by aibaars** [comment]
Replaced by #14484 and #14510
